azkaban-execserver/src/main/java/azkaban/execapp/StatisticsServlet.java

```diff
@@ -0,0 +1,284 @@
+        try {
+          java.io.BufferedReader reader =
+              new java.io.BufferedReader(new InputStreamReader(inputStream));
+          String line = null;
+          while ((line = reader.readLine()) != null) {
+            output.add(line);
+          }
+        }finally {
+          inputStream.close();
+        }
+        // process the output from bash call.
+        if (output.size() > 0) {
+          long totalMemory = 0 ;
+          long freeMemory  = 0 ;
+          int  sampleCount = 0 ;
+
+          // process all the output, we will do 5 samples.
+          for(String line : output){
+            String[] splitedresult = line.split("\\s+");
+            // expected return format -
+            // "Mem:" | total | used | free | shared | buffers | cached
+            if (splitedresult.length == 7){
+              // create a temp copy of all the readings, if anything goes wrong, we drop the
+              // temp reading and move on.
+              long tmp_totalMemory = 0 ;
+              long tmp_freeMemory  = 0 ;
+              try {
+                tmp_totalMemory = Long.parseLong(splitedresult[1]);
+                tmp_freeMemory = Long.parseLong(splitedresult[3]);
+
+              } catch(NumberFormatException e){
+                logger.error("skipping the unprocessable line from the output -" + line);
+                continue;
+              }
+
+              // add up the result.
+              ++sampleCount;
+              totalMemory += tmp_totalMemory ;
+              freeMemory  += tmp_freeMemory;
+            }
+          }
+
+          // set the value.
+          if (sampleCount > 0){
+            freeMemory  = freeMemory  / sampleCount;
+            totalMemory = totalMemory / sampleCount;
+            logger.info(String.format("total memory - %s , free memory - %s", totalMemory, freeMemory));
+            stats.setRemainingMemory(freeMemory);
+            stats.setRemainingMemoryPercent(totalMemory == 0? 0 :
+              ((double)freeMemory/(double)totalMemory));
+          }
+        }
+      }
+      catch (Exception ex){
+        logger.error("failed fetch system memory info " +
+                     "as exception is captured when fetching result from bash call.");
+      }
+    } else {
+        logger.error("failed fetch system memory info " +
+                     "as 'bash' or 'free' command can't be found on the current system.");
+    }
+  }
+
+  /**
+   * call the data providers to fill the returning data container for statistics data.
+   * This function refreshes the static cached copy of data in case if necessary.
+   * */
+  protected synchronized void populateStatistics(){
+    //check again before starting the work.
+    if (null == lastRefreshedTime ||
+        new Date().getTime() - lastRefreshedTime.getTime() > cacheTimeInMilliseconds){
+      final Statistics stats = new Statistics();
+
+      List<Thread> workerPool = new ArrayList<Thread>();
```

Comment:
It seems like this is more complicated than it needs to be.  How long does each one of these methods take to complete?